### ISSUE TITLE
[Discussion] Implementing Matter

### ISSUE THREAD

**Issue by lboue** [homebridge/homebridge#3228]
### Current Situation

Phillips has announced that they will update their _Hue Bridge V2_ to add compatibility with the Matter protocol in December 2022.
_Philips Hue Will Support “Matter” https://www.smarthomepoint.com/philips-hue-matter-support-new-video/_

In the same way it would be interesting to reuse the existing Homekit plugins and to be able to announce these devices in the Matter over IP network.
In fact, in this way, all the Matter compatible objects in the house can communicate together no matter the manufacturer.

### Proposed Change

Here is an example of a bridge available in the Matter SDK:

[CHIP Linux Bridge Example ](https://github.com/project-chip/connectedhomeip/tree/master/examples/bridge-app/linux)
An example demonstrating a simple lighting bridge and the use of dynamic endpoints. The document will describe the theory of operation and how to build and run CHIP Linux Bridge Example on Raspberry Pi. This doc is tested on Ubuntu for Raspberry Pi Server 20.04 LTS (aarch64) and Ubuntu for Raspberry Pi Desktop 20.10 (aarch64)

Do you think that would be feasible?

You can install a "Matter profile for IOS 15 and 16 to add Matter compatibily.

**Comment by MoTechnicalities** [comment]
This is a great idea. Except that most plugins are created by individuals, many being incomplete and needing a lot of work. Still, generally speaking, it would be nice to be able to bring products like Ring and WiZ into Matter if they don't intend to introduce in their own bridges.

**Comment by Neirth** [comment]
Personally, I think it is a very good opportunity to explore the option of adding native Matter support to the project, this would de facto make Homebridge one of the most complete device hubs on the market, with out-of-box support for devices that we were previously having enough problems in porting and/or maintaining, the different confirmations are happening throughout these days by the brands.  I imagine that this will be the trend until the official launch of the protocol.  

On the other hand, the great voice assistants have also confirmed support for Matter, so it will greatly simplify the work involved in integrating these technologies through plugins to the project.

TD;LR: Apple confirmed a while ago that if the device in question was compatible with Homekit it would be compatible with Matter, I understand that soon they will present us with updates in the proprietary protocol of Homekit, perhaps making it somewhat more open due to requirements of the standard .  It's just a hypothesis.

**Comment by lboue** [comment]
For information, a new home-assistant lib is in development:
[Python Matter Server](https://github.com/home-assistant-libs/python-matter-server). This project implements a Matter (formerly Home Assistant Connected Home over IP or CHIP) Controller Server using WebSockets.

**Comment by NorthernMan54** [comment]
For reference current Homebridge Architecture

```
               Home App
                   |
               HomeKit HAP Interface ( HomeKit Accessory Protocol )
                   |
              -----------------
              | HAP-NodeJS    |
              -----------------
              | Homebridge    |
              -----------------
              | Plugin        |
              -----------------
                   |
                Vendor Device API
                   |
                Vendor Device
```

For what we are discussing, adding the Matter interface in addition to the existing HAP Interface would be a change to HAP-NodeJS and not Homebridge.

**Comment by Neirth** [comment]
> I'm not sure about that. Matter and HomeKit over BLE are 2 different protocols

@lboue, I attach the developer page of Apple telling about the HomeKit/Matter support: [https://developer.apple.com/apple-home/matter/](https://developer.apple.com/apple-home/matter/). I hope this can solve your doubts about what I mentioned above.

**Comment by adriancable** [comment]
@Neirth - I think your interpretation of that page isn't what Apple intended.

As said earlier, HomeKit and Matter are separate, unrelated protocols. It is not true what you state that "if the device in question was compatible with Homekit it would be compatible with Matter". The article you link to states that the iOS HomeKit API (i.e. the 'HomeKit framework') is being extended to work transparently with Matter devices. This means, for example, that an iOS app that already implements HomeKit support won't need (many) changes to also work with Matter devices. But, this doesn't have any bearing whatsoever on the device side (e.g. Homebridge/HAP-NodeJS). Devices do not have any awareness of, any connection with, or use the iOS HomeKit framework.

HAP-NodeJS would need a complete rewrite to support Matter. It would also need to continue to support HomeKit in parallel to avoid a major feature downgrade (since a lot of HomeKit e.g. cameras are not supported at all in Matter). Not impossible, but probably (as a 'spare time project') not coming soon.

**Comment by donavanbecker** [comment]
I don't think we would want to rewrite HAP-NodeJs, but would want to give an option of using HAP-NodeJS or [Matter](https://github.com/project-chip/matter.js) within Homebridge.

If someone wants to give it a try and implement it into Homebridge, we could look at merging it into Homebridge.

**Comment by Apollon77** [comment]
Hi,

I'm one of the matter.js developers. We are in the middle of a bigger code merge and cleanups and also implement a High level API right now ... so we might not be ready now (but "soon") for such a topic and we also miss some Cluster definitions (aka Services in homekit), so if someone want to think about that we also search for contributors :-)

In fact Clusters (aka Homekit Services) are a bit different then Homekit, so mid term the question in my eyes is if not a "matterbridge" project would make more sense that uses comparable plugin structure like homebridge and bases e.g. on matter-node.js (our nodejs package that uses matter.js for putre JS basics). Then plugin devs could - with affordable efforts - recreate their plugins for the Matter structure ... but honestly I personally do not really see a 100% mapped approach. (But maybe I'm too deep in to see the options ... ) ;-)

Matter uses beside Cluster attributes (aka Characteristics) also a lot of "commands" that then execute logic (e.g "dim the light within 5seconds with effect X" and such) what homekit do not really have.

I also started a Discord at https://discord.gg/ujmRNrhDuW for "Matter integrators". So if someone wants to discuss or know more. Join there and contact me.

Ingo

**Comment by NebzHB** [comment]
Why not having a plugin like homebridge-matter that works like homebridge-gsh or homebridge-alexa to send everything into matter ? that would make sense and re-uses all the effort the community set in the plugins

**Comment by draco1544** [comment]
Any news on that ?

**Comment by jelockwood** [comment]
👍 
I am not a developer and likely not as familiar with the technicalities of both HomeBridge and Matter but I certainly support the idea of either making HomeBridge able to expose devices to Matter i.e. act as a Matter bridge, or the creation of a related separate project to accomplish this.

I am aware that at this time not all capabilities currently available via HomeKit are available via Matter e.g. apparently energy monitoring is not yet supported by Matter, so from this perspective exposing to _both_ HomeKit and Matter would give users a choice. However my own goal is to over time to switch to Matter only either natively or via Matter bridges. I am therefore currently experimenting with the Philips Hue bridge which can as both a HomeKit bridge and a Matter bridge to test lighting solutions. (There are apparently zero Matter compatible dimmer solutions for the European market as of yet.)

**Comment by afharo** [comment]
As a [plugin](https://github.com/homebridge-xiaomi-roborock-vacuum/homebridge-xiaomi-roborock-vacuum) developer, I'd love to learn what plans Apple has for device types introduced by Matter that are not currently supported in HomeKit.

In the beginning, HomeKit supported a wider variety of devices than Matter 1.0. However, Matter 1.2 adds support for robot vacuum cleaners. Will HomeKit add that [Accessory Category Type](https://developer.apple.com/documentation/homekit/hmaccessorycategory/accessory_category_types)? Or will they just direct devs to the Matter SDK (as they seem to do [in their docs](https://developer.apple.com/apple-home/matter/))?
> Xcode includes the certified Matter SDK in its entirety. You are encouraged to use this version of the SDK provided by Apple for your Matter ecosystem apps. Apple will work continuously to keep it up to date.

If the latter, I would like Homebridge to expose the extra APIs to register the matter device.

If the decision is to address it as a separate project (_matterbridge_), my [plugin](https://github.com/homebridge-xiaomi-roborock-vacuum/homebridge-xiaomi-roborock-vacuum) will need to be deprecated in favor of the matter clone (I assume many other plugins "impersonating" other accessory types in the absence of their own would be in the same situation).

TBH, I've already assumed I'll need a good rewrite of my plugin when the support for the appropriate device type comes out. However, allowing this transition by letting the plugin call different APIs based on config will be smoother for devs and users, IMO.

It could be a new set of APIs:
* `api.registerMatterAccessory` instead of `api.registerAccessory`
* `api.matter` instead of `api.hap`

IMO, that'd ease the transition for both:
* Devs: mostly the same code, just different "last mile" conversion logic
* Users: anyone wanting to test the "experimental" matter device, they can do that with a config change (maybe even keeping the accessory for their BWC automations). I'm thinking of a config like below

```diff
{
  ...
  "accessories": [
    {
      "accessory": "XiaomiRoborockVacuum",
      "name": "My Robot Vacuum",
      "ip": "192.168.0.10",
      "token": "my-secret-token"
    }
  ],
  "platforms": [...],
+  "matterAccessories": [
+    {
+      "accessory": "XiaomiRoborockVacuum",
+      "name": "My Matter Robot Vacuum",
+      "ip": "192.168.0.10",
+      "token": "my-secret-token"
+    }
+  ]
}
```

This can also be achieved by forking to project `matterbridge`, but, IMO, the effort is larger on both ends:
* Devs: Clone and publish the plugin with a new name (and try to keep feature parity during the transition, if they want to maintain any level of BWC).
* Users: Set up a new architecture altogether: new project, new NPM installs, await new UI plugin to handle the plugins, ... you know the drill.

FWIW, I don't think `homebridge` should automatically try to automatically convert compatible accessories to Matter. At least not for now. I'd rather keep it on the plugin-dev's end to implement it or not, and let the users decide whether they want to enable that mode or not for that plugin.

**Comment by nitaybz** [comment]
**Suggestion (might be an easy solution):**
There is a Matter bridge repository that exposes different plugins to matter: [https://github.com/Luligu/matterbridge](https://github.com/Luligu/matterbridge). It might be easy enough to create a matterbridge plugin that connects to Homebridge and expose all devices to matter, it can use Homebridge API or a dedicated plugin also on Homebridge side in order to connect the 2 systems.

Checkout the Zigbee2MQTT plugin for matterbridge, they connected all the accessories from Z2M into matter and it was pretty straight forward to set it up: [https://github.com/Luligu/matterbridge-zigbee2mqtt](https://github.com/Luligu/matterbridge-zigbee2mqtt)

I hope it will help to push this forward. I've created a few Homebridge plugins that I really wish to be able to expose as Matter accessories.

**Comment by smitty078** [comment]
My experience with a mix of both HomeKit (native), Matter (native), and HomeKit (via homebridge) accessories is that Matter is by far the worst experience. Matter accessories have been downright unresponsive and unreliable (both WiFi and Thread). HomeKit native and Homebridge accessories are virtual indistinguishable from each other and are generally fast and reliable in terms of connectivity. I'm not certain I want a matterbridge to exist, although conceptually I think it's a great idea, and it future-proofs my setup for the day when HomeKit inevitably becomes unsupported (someday we'll also need a HomeKit native -> Matter bridge, but that's another topic and I believe the tools to do so already exist).

It's my opinion that effectively what needs to happen is:

Option 1:
A HAP-NodeJS equivalent for Matter library to exist

Option 2:
Matter support is added to HAP-NodeJS

With either option:
homebridge needs to implement using both new/updated libraries

**Comment by adriancable** [comment]
@smitty078 - whether you want a matterbridge to exist or not, it does: https://github.com/Luligu/matterbridge - this is more or less equivalent to Homebridge for HomeKit, i.e. it's a way for developers to create plug-ins that publish Matter devices.

Your option 1 also already exists, it's matter.js which is the official NodeJS implementation of the Matter device SDK, which matterbridge uses.

While no tech products exist forever, HomeKit will continue to be supported for a very long time. Smart home devices are not like phones, they have very long replacement cycles (thermostats typically go decades between replacement), most HomeKit devices in existence cannot be upgraded (or are not commercially viable to upgrade) to support Matter, and Apple are aware that it is a space where, once you are in, you can basically never get out.

**Comment by paolotremadio** [comment]
> My experience with a mix of both HomeKit (native), Matter (native), and HomeKit (via homebridge) accessories is that Matter is by far the worst experience. Matter accessories have been downright unresponsive and unreliable (both WiFi and Thread). HomeKit native and Homebridge accessories are virtual indistinguishable from each other and are generally fast and reliable in terms of connectivity. I'm not certain I want a matterbridge to exist, although conceptually I think it's a great idea, and it future-proofs my setup for the day when HomeKit inevitably becomes unsupported (someday we'll also need a HomeKit native -> Matter bridge, but that's another topic and I believe the tools to do so already exist).

I have a few Matter accessories and they've been super responsive in Home Assistant but absolutely terrible in Apple Home. I'm starting to suspect there's something wrong in the Apple Home implementation. But I do agree with you: HAP accessories have always been very stable.


> Option 1: A HAP-NodeJS equivalent for Matter library to exist

As already pointed out, that exists already. Matter is now providing an official Javascript/Typescript implementation and it works well. Documentation needs some improvement but it works well.


> Option 2: Matter support is added to HAP-NodeJS

That is not as straight forward as it looks. The reason is due to how Matter works (which is very similar to how Zigbee works).

In HAP you have accessories exposing characteristics. In Matter you have nodes exposing clusters with attributes. Pretty much the same thing but with different naming.

In HAP you can read, write or subscribe to a characteristic. This is where Matter diverges quite a bit. For basic bits you can read/write/subscribe attributes (e.g. current temperature / target temperature on a thermostat). What is different in Matter is _commands_. Lets say you want to change brightness of a lightbulb; in HAP you set brightness characteristic to the desired value. In Matter you execute a command like `MoveToLevel` where you can specify other attributes like transition time. This is why on Philips Hue you can fade the lights when changing brightness but you cannot do that in HAP.

Another examples of where commands are useful is Locking/Unlocking a lock mechanism. In HAP you just write the desired target value. In Matter you run a lock command with a timeout (I won't explain in details how it works but it does improve security) or you can add a PIN code to the unlock command.

Another bit that is quite useful of Matter is the ability of adding "metadata" which is not strictly functional but it helps the controller displaying a more appropriate UI (e.g. you can say what colour the window covering is or how the buttons are arrange on a controller). HAP does have some support for some elements but not all of those.

My point is: you can't just translate an homebridge plugin to a Matter one without a lot of approximation and some very opinionated defaults.

**Comment by Pythonaire** [comment]
For me, it looks like i had expected. I made my own sensor devices and connected them to the HomeKit based on the HAP public build for services and characteristics - so few losses in the UI. Device manufacturer want to sell devices and Apple for example, want to sell his eco system (HomeKit, iPhone etc.), the result is the MFI Program or additionally bridges, separate UI apps etc. Maybe Matter will help to share data between the top dogs API's of google, apple and amazon in this field. But none of them are really interested in that. Using metadatas,  instead of special service and characteristic dataset is truly a good solution, anybody can handle that, but at least the API, however it calls, must handle these metadata. If i made i sensor for measuring the moon cycles, no-one is able to represent that, with or without Matter, unless i pay for it.

**Comment by justjam2013** [comment]
Would something like this be feasible?

```
                           Home App
                     _________|________
                     |                |
                HomeKit HAP         Matter
                     |                |
              ----------------------------------
              | HAP-NodeJS    |    Matter.js   |
              ----------------------------------
              |           Homebridge           |
              ----------------------------------
              |  HAP Plugin  |  Matter Plugin  |
              ----------------------------------
                        |_________|
                             |
                     Vendor Device API
                             |
                       Vendor Device
```

Instead of abstracting or bridging across the protocols, Homebridge simply exposes the necessary classes from Matter.js.
Similarly to `Service` and `Characteristic` which are defined in the `hap-nodejs` node-module and exposed by the `homebridge` node-module.

Homebridge retains its role of managing the plugins without caring about the protocol they use. I will say up front that I don't know enough about Homebridge to be able to determine how much effort this would be.

However, the idea is that Homebridge would support HAP plugins and Matter plugins, and leave it up to plugin developers to decide which protocol to use.

**Comment by Pythonaire** [comment]
You are right. But what is the advantage? Try to imagine a conversation with a alien. The API or the protocol is the translator. 
Matter can translate the term "car" from different languages, but if the alien don't know what a car is, it make not sense, no matter how may translator you have. The alien "apple homekit" knows bicicles, airplanes and other transportation devices, but not cars. You can say "car" in any languages and translate that in any protocol - the conversation will not work.

**Comment by thoukydides** [comment]
I can see compelling arguments both for and against Homebridge integrating Matter support.

In favour:
- It provides a simpler migration from HAP to Matter based plugins for end-users. Other than having to pair the bridge once as HAP and once as Matter, everything else is transparent. The configuration and management interface is unchanged.
- Plugins can mix-and-match use of HAP accessories and Matter devices, whether to best utilise the Home app’s capabilities, or to allow incremental migration. (For example, my Homebridge-homeconnect plugin could start by implementing the robot vacuum cleaners as Matter RVCs leaving the kitchen appliances as HAP until the Home app implements the corresponding Matter clusters.)
- If a mixture of HAP and Matter is required for a particular plugin then it avoids problems with API rate limits, i.e. it doesn’t require two plugins polling the same API. This is a serious issue for many of the cloud-based APIs that I use - both in terms of their hard rate limits, and the required try-again-after behaviour.

Against:
- HAP and Matter are very different. The interfaces and plugin behaviour for each will be almost entirely separate. A plugin switching from one to the other will replace most of that code, probably just keeping their device-specific API implementation. A plugin supporting both simultaneously will have two largely independent protocol conversion stacks.
- Homebridge would become significantly more complicated, and its maintainers would need to become Matter experts.
- Support for Matter is still immature, both in clients (like the Home app) and server libraries (like matter.js). Homebridge is now mature and stable, but introducing Matter would destabilise it.
- Matter is supported by clients other than HomeKit. Plugins bridging to Matter would be useful for non-Apple (or mixed ecosystem) users, but implementing them in Homebridge would largely prevent non-Apple users from using them.

My personal preference would be for Homebridge to integrate Matter support. However, I am currently porting some of my plugins to Matterbridge…

**Comment by Losalado** [comment]
Any update for implementing Matter on HomeBridge?

**Comment by pizzapi2012** [comment]
https://github.com/homebridge/homebridge/commit/a2287c54857d1b7e3dfb2a7e51661d34b36bd708 👀

**Comment by pizzapi2012** [comment]
they are working on it

**Comment by NorthernMan54** [comment]
@pizzapi2012 The Alpha release stream contains the WIP.  Still early and not truly functional.  You can try the Alpha stream via the Homebridge Docker and VM Images beside your current setup if you want to explore.

**Comment by pizzapi2012** [comment]
> [@pizzapi2012](https://github.com/pizzapi2012) The Alpha release stream contains the WIP. Still early and not truly functional. You can try the Alpha stream via the Homebridge Docker and VM Images beside your current setup if you want to explore.

ok, and now it is in beta

**Comment by donavanbecker** [comment]
> ok, and now it is in beta

There is even a homebridge-matter plugin if you want to try out virtual devices

**Comment by a-crazy-pencil** [comment]
Should this issue be closed? They already added it

**Comment by NorthernMan54** [comment]
And implemented with HB 2.0
